### ISSUE TITLE
BFLATN: store sum tags before payloads, same as BSATN

### ISSUE THREAD

**Issue by gefjon** [clockworklabs/SpacetimeDB#1006]
This would improve the fast-path BFLATN -> BSATN conversion, as it would allow larger `memcpy` blocks.

**Comment by kazimuth** [comment]
I'm concerned that even with this change memcpy still won't be possible for most sum types, because BSATN does not take alignment into account.

e.g.
```rust
enum Thing {
     Stuff(u32)
}
```

bsatn layout:
| `0x0` | `0x1`  | `0x2`  | `0x3`  | `0x4`  |
| ----- | ------ | ------ | ------ | ------ |
| tag   | stuff0 | stuff1 | stuff2 | stuff3 |

old bflatn layout:
| `0x0`  | `0x1`  | `0x2`  | `0x3`  | `0x4` |
| ------ | ------ | ------ | ------ | ----- |
| stuff0 | stuff1 | stuff2 | stuff3 | tag   |

new bflatn layout (taking into account that `u32` has a required alignment of 4)
| `0x0` | `0x1` | `0x2` | `0x3` | `0x4`  | `0x5`  | `0x6`  | `0x7` |
| ----- | ----- | ----- | ----- | ------ | ------ | ------ | ------ |
| tag   | pad   | pad   | pad   | stuff0 | stuff1 | stuff2 | stuff3 |

**Comment by gefjon** [comment]
You are correct, and we don't want to introduce padding into BSATN. The reason swapping the order would be beneficial is that, for enums where the fast-path does apply, we could merge copying the tag and/or the payload with preceeding and/or succeeding data.

**Comment by kazimuth** [comment]
ahhh right okay cool.
